### PR TITLE
Move model-related fields in a subfield for the Python Predictor

### PR DESCRIPTION
**Pull request by RobertLucian**
Closes cortexlabs/internal#20, closes cortexlabs/internal#39.

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [x] update examples
- [x] update docs and add any new files to `summary.md` (view in [gitbook](https://docs.cortex.dev/v/master) after merging)

### CODE REVIEW

File: test/live-reloading/python/mpg-estimator/cortex.yaml [hunk 1]
```diff
@@ -5,4 +5,5 @@
   predictor:
     type: python
     path: predictor.py
-    model_path: s3://cortex-examples/sklearn/mpg-estimator/linreg/
+    dynamic_model_loading:
+      model_path: s3://cortex-examples/sklearn/mpg-estimator/linreg/
```
Comment (RobertLucian) [inline test/live-reloading/python/mpg-estimator/cortex.yaml:9]: This needs to be changed to an iris-classifier example that has at least 3 versions.

File: test/model-caching/python/mpg-estimator/cortex.yaml [hunk 2]
```diff
@@ -5,9 +5,10 @@
...
-      paths:
-        - name: mpg-estimator
-          model_path: s3://cortex-examples/sklearn/mpg-estimator/linreg/
-      cache_size: 1
-      disk_cache_size: 2
+    dynamic_model_loading:
+      models:
+        paths:
+          - name: mpg-estimator
+            model_path: s3://cortex-examples/sklearn/mpg-estimator/linreg/
```
Comment (RobertLucian) [inline test/model-caching/python/mpg-estimator/cortex.yaml:12]: This needs to be changed to an iris-classifier example that has at least 3 versions.

File: docs/workloads/realtime/configuration.md [hunk 3]
```diff
@@ -9,12 +9,12 @@
   predictor:
     type: python
     path: <string>  # path to a python file with a PythonPredictor class definition, relative to the Cortex root (required)
-    model_path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (optional, cannot be provided along with 'models')
-    models:  # use this to serve multiple models in a single API (optional, cannot be provided along with 'model_path')
-      dir: <string>  # S3 path to a directory containing multiple models (e.g. s3://my-bucket/models/) (either this or 'paths' must be provided)
-      paths:  # list of S3 paths to exported model directories (either this or 'dir' must be provided)
+    multi_model_reloading:  # use this to serve a single model or multiple ones (optional)
+      dir: <string>  # S3 path to a directory containing multiple models (e.g. s3://my-bucket/models/)
```
Comment (deliahu) [inline docs/workloads/realtime/configuration.md:13]: (either this, 'path', or 'paths' must be provided)

Comment (deliahu) [inline docs/workloads/realtime/configuration.md:13]: nit: move dir after paths

File: docs/workloads/realtime/configuration.md [hunk 4]
```diff
@@ -66,14 +66,13 @@
   predictor:
     type: tensorflow
     path: <string>  # path to a python file with a TensorFlowPredictor class definition, relative to the Cortex root (required)
-    model_path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (either this or 'models' must be provided)
-    signature_key: <string>  # name of the signature def to use for prediction (required if your model has more than one signature def)
-    models:  # use this to serve multiple models in a single API (either this or 'model_path' must be provided)
-      dir: <string>  # S3 path to a directory containing multiple models (e.g. s3://my-bucket/models/) (either this or 'paths' must be provided)
-      paths:  # list of S3 paths to exported model directories (either this or 'dir' must be provided)
+    models:  # use this to serve a single model or multiple ones
```
Comment (deliahu) [inline docs/workloads/realtime/configuration.md:69]: add "(required)" to end of comment

File: docs/workloads/realtime/configuration.md [hunk 5]
```diff
@@ -66,14 +66,13 @@
...
+      dir: <string>  # S3 path to a directory containing multiple models (e.g. s3://my-bucket/models/)
+      path: <string> # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/)
+      paths:  # list of S3 paths to exported model directories
         - name: <string>  # unique name for the model (e.g. text-generator) (required)
-          model_path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (required)
-          signature_key: <string>  # name of the signature def to use for prediction (required if your model has more than one signature def)
+          path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (required)
+          signature_key: <string>  # name of the signature def to use for prediction (optional)
         ...
       signature_key:  # name of the signature def to use for prediction for 'dir'-specified models or for models specified using 'paths' that haven't had a signature key set
```
Comment (deliahu) [inline docs/workloads/realtime/configuration.md:77]: nit: `name of the signature def to use for prediction (required if your model has more than one signature def)`

File: docs/workloads/realtime/configuration.md [hunk 6]
```diff
@@ -66,14 +66,13 @@
...
-      paths:  # list of S3 paths to exported model directories (either this or 'dir' must be provided)
+    models:  # use this to serve a single model or multiple ones
+      dir: <string>  # S3 path to a directory containing multiple models (e.g. s3://my-bucket/models/)
+      path: <string> # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/)
+      paths:  # list of S3 paths to exported model directories
         - name: <string>  # unique name for the model (e.g. text-generator) (required)
-          model_path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (required)
-          signature_key: <string>  # name of the signature def to use for prediction (required if your model has more than one signature def)
+          path: <string>  # S3 path to an exported model directory (e.g. s3://my-bucket/exported_model/) (required)
+          signature_key: <string>  # name of the signature def to use for prediction (optional)
```
Comment (deliahu) [inline docs/workloads/realtime/configuration.md:75]: nit: `name of the signature def to use for prediction (required if your model has more than one signature def)`

File: docs/workloads/realtime/models.md [hunk 7]
```diff
@@ -168,7 +170,7 @@ In this case, there are two models in the directory, one of which is named "text
 
 ## Live model reloading
 
-Live model reloading is a mechanism that periodically checks for updated models in the model path(s) provided in `predictor.model_path` or `predictor.models`. It is automatically enabled for all predictor types, including the Python predictor type (as long as model paths are specified via `model_path` or `models` in the `predictor` configuration).
+Live model reloading is a mechanism that periodically checks for updated models in the model path(s) provided in `predictor.models`. It is automatically enabled for all predictor types, including the Python predictor type (as long as model paths are specified via `models` in the `predictor` configuration).
```
Comment (deliahu) [inline docs/workloads/realtime/models.md:173]:
```
specified via `models` -> specified via `multi_model_reloading`
```

File: docs/workloads/realtime/predictors.md [hunk 8]
```diff
@@ -59,8 +59,7 @@ class PythonPredictor:
                 the model and/or metadata.
             python_client (optional): Python client which is used to retrieve
                 models for prediction. This should be saved for use in predict().
-                Required when `predictor.model_path` or `predictor.models` is
-                specified in the api configuration.
+                Required when `predictor.models` is specified in the api configuration.
```
Comment (deliahu) [inline docs/workloads/realtime/predictors.md:62]:
```
`predictor.models` -> `predictor.multi_model_reloading`
```

File: docs/workloads/realtime/predictors.md [hunk 9]
```diff
@@ -103,7 +102,7 @@ class PythonPredictor:
     def load_model(self, model_path):
         """(Optional) Called by Cortex to load a model when necessary.
 
-        This method is required when `predictor.model_path` or `predictor.models`
+        This method is required when `predictor.models`
```
Comment (deliahu) [inline docs/workloads/realtime/predictors.md:105]:
```
`predictor.models` -> `predictor.multi_model_reloading`
```

File: pkg/operator/operator/k8s.go [hunk 10]
```diff
@@ -410,7 +410,15 @@ func getEnvVars(api *spec.API, container string) []kcore.EnvVar {
 			)
 		}
 
-		if api.Predictor.ModelPath != nil || api.Predictor.Models != nil {
+		if api.Predictor.Type == userconfig.PythonPredictorType && api.Predictor.MultiModelReloading != nil {
```
Comment (deliahu) [inline pkg/operator/operator/k8s.go:413]: consider: `if api.Predictor.Type != userconfig.PythonPredictorType || api.Predictor.MultiModelReloading != nil {`

File: pkg/types/spec/validations.go [hunk 11]
```diff
@@ -817,28 +813,53 @@ func validatePredictor(
...
+		if api.Predictor.Type != userconfig.PythonPredictorType {
+			return ErrorFieldNotSupportedByPredictorType(userconfig.MultiModelReloadingKey, api.Predictor.Type)
+		}
+		models = api.Predictor.MultiModelReloading
+	}
+	if models == nil && api.Predictor.Type != userconfig.PythonPredictorType {
+		return ErrorFieldMustBeDefinedForPredictorType(userconfig.ModelsKey, api.Predictor.Type)
+	}
+
+	if models.Path == nil && len(models.Paths) == 0 && models.Dir == nil {
```
Comment (deliahu) [inline pkg/types/spec/validations.go:833]: Would this panic for python predictor which does not specify models?

File: pkg/types/spec/validations.go [hunk 12]
```diff
@@ -817,28 +813,53 @@ func validatePredictor(
...
+		return errors.Wrap(ErrorSpecifyOnlyOneField(userconfig.ModelsPathKey, userconfig.ModelsPathsKey, userconfig.ModelsDirKey), userconfig.ModelsKey)
+	}
+	if models.Path != nil && (len(models.Paths) > 0 || models.Dir != nil) {
```
Comment (deliahu) [inline pkg/types/spec/validations.go:836]: I think this should be three ifs, e.g.

`if models.Path != nil && len(models.Paths) > 0 {`

`if models.Dir != nil && len(models.Paths) > 0 {`

`if models.Dir != nil && models.Path != nil {`

Comment (deliahu) [reply pkg/types/spec/validations.go:836]: Consider adding a check if they are all specified

File: pkg/types/spec/validations.go [hunk 13]
```diff
@@ -867,60 +889,69 @@ func validatePythonPredictor(api *userconfig.API, models *[]CuratedModelResource
 		return ErrorFieldNotSupportedByPredictorType(userconfig.TensorFlowServingImageKey, predictor.Type)
 	}
 
-	hasSingleModel := predictor.ModelPath != nil
-	hasMultiModels := predictor.Models != nil
+	if predictor.MultiModelReloading == nil {
+		return nil
+	}
+	dml := predictor.MultiModelReloading
```
Comment (deliahu) [inline pkg/types/spec/validations.go:895]: nit: dml -> mmr

File: pkg/types/userconfig/api.go [hunk 14]
```diff
@@ -647,31 +643,42 @@ func (api *API) TelemetryEvent(provider types.ProviderType) map[string]interface
...
+		}
+		if api.Predictor.MultiModelReloading != nil {
+			models = api.Predictor.MultiModelReloading
+		}
+
+		if models != nil {
 			event["predictor.models._is_defined"] = true
-			if len(api.Predictor.Models.Paths) > 0 {
+			if models.Path != nil {
+				event["predictor.models.model_path._is_defined"] = true
```
Comment (deliahu) [inline pkg/types/userconfig/api.go:657]: model_path -> path

File: pkg/workloads/cortex/lib/model/type.py [hunk 15]
```diff
@@ -147,84 +147,88 @@ def get_models_from_api_spec(
     api_spec: dict, model_dir: str = "/mnt/model"
 ) -> CuratedModelResources:
     """
-    Only effective for predictor:model_path, predictor:models:paths or for predictor:models:dir when the dir is a local path.
-    It does not apply for when predictor:models:dir is set to an S3 model path.
+    Only effective for models.path, models:paths or for models:dir fields when the dir is a local path.
```
Comment (deliahu) [inline pkg/workloads/cortex/lib/model/type.py:150]: nit: semicolon notation

**Review by deliahu — APPROVED** [review]

File: pkg/types/spec/validations.go (hunk 11 above)
Comment (RobertLucian) [reply pkg/types/spec/validations.go:941]: Yes, this panics. It is now addressed.

File: pkg/types/spec/validations.go (hunk 12 above)
Comment (RobertLucian) [reply pkg/types/spec/validations.go:944]: Added the check if they are all specified as well.

**Pull request merged**